File: app/src/main/java/org/sil/storyproducer/media/pipe/PipedAudioLooper.java
package org.sil.storyproducer.media.pipe;

import android.media.MediaCodec;
import android.media.MediaFormat;

import org.sil.storyproducer.media.MediaHelper;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ShortBuffer;

/**
 * <p>This media pipeline component loops a single audio file for a specified amount of time.</p>
 */
public class PipedAudioLooper extends PipedAudioShortManipulator {
    private static final String TAG = "PipedAudioMixer";

    private final String mPath;
    private final long mDurationUs;

    private float mVolumeModifier = 1f;

    private PipedMediaByteBufferSource mSource;

    public PipedAudioLooper(String src, long durationUs) {
        mPath = src;
        mDurationUs = durationUs;
    }

    public PipedAudioLooper(String src, long durationUs, int sampleRate, int channelCount) {
        this(src, durationUs, sampleRate, channelCount, 1);
    }

    public PipedAudioLooper(String src, long durationUs, int sampleRate, int channelCount, float volumeModifier) {
        mPath = src;
        mDurationUs = durationUs;
        mSampleRate = sampleRate;
        mChannelCount = channelCount;
        mVolumeModifier = volumeModifier;
    }

    private MediaFormat mOutputFormat;

    private ByteBuffer mSourceBuffer;
    private ShortBuffer mSourceShortBuffer;

    private MediaCodec.BufferInfo mInfo = new MediaCodec.BufferInfo();

    @Override
    public void setup() throws IOException, SourceUnacceptableException {
        mSource = new PipedAudioDecoderMaverick(mPath, mSampleRate, mChannelCount, mVolumeModifier);
        mSource.setup();
        fetchSourceBuffer();

        mOutputFormat = MediaHelper.createFormat(MediaHelper.MIMETYPE_RAW_AUDIO);
        mOutputFormat.setInteger(MediaFormat.KEY_SAMPLE_RATE, mSampleRate);
        mOutputFormat.setInteger(MediaFormat.KEY_CHANNEL_COUNT, mChannelCount);
    }

    @Override
    public MediaFormat getOutputFormat() {
        return mOutputFormat;
    }

    @Override
    public boolean isDone() {
        return mSeekTime >= mDurationUs;
    }

    @Override
    protected short getSampleForTime(long time, int channel) {
        while(mSourceShortBuffer != null && mSourceShortBuffer.remaining() <= 0) {
            releaseSourceBuffer();
            fetchSourceBuffer();
        }
        if(mSourceShortBuffer == null) {
            mSource.close();
            mSource = new PipedAudioDecoderMaverick(mPath, mSampleRate, mChannelCount, mVolumeModifier);
            try {
                mSource.setup();
                fetchSourceBuffer();
            } catch (IOException e) {
                //TODO: handle exception properly
                e.printStackTrace();
            } catch (SourceUnacceptableException e) {
                //TODO: handle exception properly
                e.printStackTrace();
            }
        }
        if(mSourceShortBuffer != null) {
            return mSourceShortBuffer.get();
        }

        //TODO: can this happen?
        return 0;
    }

    private void fetchSourceBuffer() {
        if(mSource.isDone()) {
            return;
        }
        ByteBuffer buffer = mSource.getBuffer(mInfo);
        mSourceBuffer = buffer;
        mSourceShortBuffer = MediaHelper.getShortBuffer(buffer);
    }

    private void releaseSourceBuffer() {
        mSource.releaseBuffer(mSourceBuffer);
        mSourceBuffer = null;
        mSourceShortBuffer = null;
    }

    @Override
    public void close() {
        super.close();
        if(mSource != null) {
            mSource.close();
        }
    }
}

File: app/src/main/java/org/tyndalebt/spadv/tools/media/pipe/PipedAudioLooper.kt
package org.tyndalebt.spadv.tools.media.pipe

import android.content.Context
import android.media.MediaFormat
import android.util.Log
import org.tyndalebt.spadv.tools.media.MediaHelper
import java.io.IOException

/**
 *
 * This media pipeline component loops a single audio file for a specified amount of time.
 */
class PipedAudioLooper
/**
 * Create looper from an audio file with specified duration, resampling the audio stream.
 * @param mPath path of the audio file.
 * @param mDurationUs desired duration in microseconds.
 * @param sampleRate desired sample rate.
 * @param channelCount desired channel count.
 * @param mVolumeModifier volume scaling factor.
 */
@JvmOverloads constructor(private val context: Context, private val mPath: String, private val mDurationUs: Long, sampleRate: Int = 0, channelCount: Int = 0, private val mVolumeModifier: Float = 1f) : PipedAudioShortManipulator() {

    private var mOutputFormat: MediaFormat? = null

    override val componentName: String = TAG

    init {
        mSampleRate = sampleRate
        mChannelCount = channelCount
    }

    override fun getOutputFormat(): MediaFormat? {
        return mOutputFormat
    }

    @Throws(IOException::class, SourceUnacceptableException::class)
    override fun setup() {
        if (mComponentState != PipedMediaSource.State.UNINITIALIZED) {
            return
        }

        mSource = PipedAudioDecoderMaverick(context, mPath, mSampleRate, mChannelCount, mVolumeModifier)
        mSource!!.setup()

        validateSource(mSource!!)

        try {
            fetchSourceBuffer()
        } catch (e: SourceClosedException) {
            //This case should not happen.
            throw SourceUnacceptableException("First fetchSourceBuffer failed! Strange", e)
        }

        val sourceOutputFormat = mSource!!.outputFormat
        mSampleRate = sourceOutputFormat.getInteger(MediaFormat.KEY_SAMPLE_RATE)
        mChannelCount = sourceOutputFormat.getInteger(MediaFormat.KEY_CHANNEL_COUNT)

        mOutputFormat = MediaHelper.createFormat(MediaHelper.MIMETYPE_RAW_AUDIO)
        mOutputFormat!!.setInteger(MediaFormat.KEY_SAMPLE_RATE, mSampleRate)
        mOutputFormat!!.setInteger(MediaFormat.KEY_CHANNEL_COUNT, mChannelCount)

        mComponentState = PipedMediaSource.State.SETUP

        start()
    }

    @Throws(SourceClosedException::class)
    override fun loadSamples(): Boolean {
        //Component is done if duration is exceeded.
        if (mSeekTime >= mDurationUs) {
            mSource!!.close()
            mSource = null
            return false
        }

        if (srcHasBuffer && srcPos >= srcEnd) {
            fetchSourceBuffer()
        }
        if (!srcHasBuffer) {
            mSource!!.close()
            mSource = PipedAudioDecoderMaverick(context, mPath, mSampleRate, mChannelCount, mVolumeModifier)

            try {
                mSource!!.setup()
                fetchSourceBuffer()
            } catch (e: IOException) {
                Log.e(TAG, "Source setup failed!", e)
                mSource!!.close()
                mSource = null
                return false
            } catch (e: SourceUnacceptableException) {
                Log.e(TAG, "Source setup failed!", e)
                mSource!!.close()
                mSource = null
                return false
            }

        }

        return true
    }

    companion object {
        private val TAG = "PipedAudioLooper"
    }
}